### DISCUSSION TITLE
How to reference "New Column" defined column to define new column？

### DISCUSSION THREAD

**Discussion by hengaini2055** [perspective-dev/perspective#2150] [Q&A]
### Description of Problem:
I can define a new column such as sum("Sales")  named SalesSum and use in group by datagrid. I want to define a new column reference the defined column SalesSum, the error is "Value error - input column "SalesSum" does not exists".

### Potential Solutions:
I think the problem is the "New column" function can only reference the table source column and can not  reference "New column" defined virtual column .
If we can reference virtual column defined by "New column", the perspective tool will open a big windows for data analysis.

**Comment by texodus** [comment]
We used to allow this but removed it because it introduced a lot of UX complexity.

> ... will open a big windows for data analysis.

Details about your use case would help a lot.

**Comment by hengaini2055** [comment]
If I want to display a metric as [sum(profit) / sum(Sales)],  how to accomplish this in perspective?

**Comment by timkpaine** [comment]
you need to provide more details. given columns `profit` and `sales` you would do `sum(“profit”) / sum(“sales”)`. but if your columns `profit` and `sales` are themselves computed columns, then you cannot as mentioned above.

**Comment by hengaini2055** [comment]
As you mentioned, I would do sum("profit") / sum("Sales"), but the value is wrong.
![image](https://user-images.githubusercontent.com/20587475/227696943-a9dbba58-0dd8-48f1-aa5c-3b415616f0c3.png)

**Comment by hengaini2055** [comment]
This picture looks more clear.
![image](https://user-images.githubusercontent.com/20587475/227697070-02f0ed40-31b5-4466-88ed-850205bcbf6b.png)

**Comment by texodus** [comment]
`sum` in expression syntax is not a rollup - you can't rollup within an expression and expressions are calculated before aggregates (which in this case is also `sum`, but this version _is_ a rollup, calculated last).  It is the sum of the vararg inputs, for which you are only providing one.

If we rename the aggregate `col_sum` to disambiguate, the highlighted value in your screenshot is essentially `col_sum(sum("sales", ..) / sum("profit", ...))`, which because your expression `sum` has only one argument is the identity function, so this column is now `col_sum("sales" / "profit")`.  Please review the [perspective expression docs](https://perspective.finos.org/docs/expressions/) which cover evaluation order in detail, and the function references in [perspective docs](https://perspective.finos.org/docs/obj/perspective-viewer-exprtk/#sum)  and [exprtk docs](https://github.com/ArashPartow/exprtk) (which are also visible in the function definition autocomplete dropdown within the component).

**Comment by hengaini2055** [comment]
Can we add a new calculation layer? That based on the group by or rollup dataset.  That will be simple, the new column can reference this dataset column as before. The successful method should do once more.
I am sorry, I don't know the internal mechanism of perspective; but frankly,  this requirement is so intuitive.
